File: api/src/main/java/fr/huiitre/tools/modules/riot/valorant/infrastructure/RiotAuthHttpAdapter.java
package fr.huiitre.tools.modules.riot.valorant.infrastructure;

import fr.huiitre.tools.modules.riot.valorant.application.core.ports.RiotAuthPort;
import org.springframework.core.ParameterizedTypeReference;
import org.springframework.http.*;
import org.springframework.util.LinkedMultiValueMap;
import org.springframework.util.MultiValueMap;
import org.springframework.web.client.HttpClientErrorException;
import org.springframework.web.client.RestTemplate;

import java.time.LocalDateTime;
import java.util.Map;

public class RiotAuthHttpAdapter implements RiotAuthPort {

    private static final String TOKEN_URL = "https://auth.riotgames.com/token";
    private static final String CLIENT_ID = "prod-xsso-playvalorant";

    private final RestTemplate restTemplate;
    private final ValorantTokenParser tokenParser;

    public RiotAuthHttpAdapter(RestTemplate restTemplate, ValorantTokenParser tokenParser) {
        this.restTemplate = restTemplate;
        this.tokenParser = tokenParser;
    }

    @Override
    public ValorantAuthResponse refresh(String refreshToken) {
        HttpHeaders headers = new HttpHeaders();
        headers.setContentType(MediaType.APPLICATION_FORM_URLENCODED);

        MultiValueMap<String, String> body = new LinkedMultiValueMap<>();
        body.add("grant_type", "refresh_token");
        body.add("refresh_token", refreshToken);
        body.add("client_id", CLIENT_ID);

        HttpEntity<MultiValueMap<String, String>> request = new HttpEntity<>(body, headers);

        try {
            ResponseEntity<Map<String, Object>> response = restTemplate.exchange(
                    TOKEN_URL, HttpMethod.POST, request,
                    new ParameterizedTypeReference<>() {});

            Map<String, Object> responseBody = response.getBody();
            if (responseBody == null) {
                throw new IllegalArgumentException("RIOT_AUTH_EMPTY_RESPONSE");
            }

            String accessToken = (String) responseBody.get("access_token");
            String newRefreshToken = (String) responseBody.get("refresh_token");
            
            // Utilisation de l'extracteur partagé
            String puuid = tokenParser.extractPuuid(accessToken);
            
            // Par défaut, le refresh token Riot dure 30 jours
            LocalDateTime expiresAt = LocalDateTime.now().plusDays(30);

            return new ValorantAuthResponse(accessToken, newRefreshToken, puuid, expiresAt);

        } catch (HttpClientErrorException e) {
            throw new IllegalArgumentException("RIOT_TOKEN_INVALID");
        } catch (Exception e) {
            throw new RuntimeException("Erreur technique lors du refresh Riot", e);
        }
    }
}

File: api/Modules/Riot/Valorant/Infrastructure/RiotAuthHttpAdapter.cs
using System.Net.Http.Json;
using System.Text.Json;
using Tools.Api.Modules.Core.Common.Application.Exceptions;
using Tools.Api.Modules.Riot.Valorant.Application.Core.Ports;
using Tools.Api.Modules.Riot.Valorant.Application.Core.Services;

namespace Tools.Api.Modules.Riot.Valorant.Infrastructure;

// Échange le refresh token contre un access token, sur le point d'entrée OAuth de Riot.
public sealed class RiotAuthHttpAdapter(
    HttpClient httpClient,
    IValorantTokenParser tokenParser) : IRiotAuthPort
{
    private const string TokenUrl = "https://auth.riotgames.com/token";
    private const string ClientId = "prod-xsso-playvalorant";

    // Riot ne dit pas combien de temps vit le refresh token qu'il rend. Trente jours est la durée
    // observée, reprise telle quelle du Java : la colonne expires_at n'est qu'indicative, rien ne
    // s'appuie dessus pour décider d'un renouvellement.
    private static readonly TimeSpan RefreshTokenLifetime = TimeSpan.FromDays(30);

    public async Task<IRiotAuthPort.ValorantAuthResponse> Refresh(string refreshToken)
    {
        HttpResponseMessage response;

        try
        {
            response = await httpClient.PostAsync(TokenUrl, new FormUrlEncodedContent(new Dictionary<string, string>
            {
                ["grant_type"] = "refresh_token",
                ["refresh_token"] = refreshToken,
                ["client_id"] = ClientId
            }));
        }
        catch (Exception exception) when (exception is HttpRequestException or TaskCanceledException)
        {
            throw AppException.Unavailable(
                "RIOT_AUTH_UNAVAILABLE",
                $"Le service d'authentification Riot est injoignable : {exception.Message}");
        }

        // Riot refuse un refresh token périmé par un 4xx. Ce code déclenche la suppression du
        // compte lié en amont : il ne doit pas être posé sur une panne réseau ou un 5xx.
        if ((int)response.StatusCode is >= 400 and < 500)
        {
            throw AppException.Validation(
                ValorantAuthService.TokenInvalidCode,
                "Riot a refusé ce refresh token : il faut relier le compte.");
        }

        if (!response.IsSuccessStatusCode)
        {
            throw AppException.Unavailable(
                "RIOT_AUTH_UNAVAILABLE",
                $"Le service d'authentification Riot a répondu {(int)response.StatusCode}.");
        }

        var payload = await response.Content.ReadFromJsonAsync<JsonElement>();

        var accessToken = ReadString(payload, "access_token");
        var newRefreshToken = ReadString(payload, "refresh_token");

        return new IRiotAuthPort.ValorantAuthResponse(
            accessToken,
            newRefreshToken,
            tokenParser.ExtractPuuid(accessToken),
            DateTime.UtcNow.Add(RefreshTokenLifetime));
    }

    private static string ReadString(JsonElement payload, string propertyName) =>
        payload.ValueKind == JsonValueKind.Object
        && payload.TryGetProperty(propertyName, out var property)
        && property.GetString() is { Length: > 0 } value
            ? value
            : throw AppException.Unavailable(
                "RIOT_AUTH_EMPTY_RESPONSE",
                $"La réponse de Riot ne contient pas « {propertyName} ».");
}
